### PR TITLE
Code refactoring , Executor function DRY

### PR DESCRIPTION
**Pull request by Hassan754**
Refactored the executor_function and made it reusable

### CODE REVIEW

File: d3a_api_client/redis_aggregator.py [hunk 1]
```diff
@@ -167,51 +167,31 @@ def batch_command(self, batch_command_dict, is_blocking=True):
 
     def _on_market_cycle(self, message):
         logging.info(f"A new market was created. Market information: {message}")
+        function_name = "on_market_cycle"
```
Comment (spyrostz) [inline d3a_api_client/redis_aggregator.py:170]: I would very much prefer if you avoided selecting the methods using strings, and use closures instead. It is far more readable, and still one liner, to create a method `event_closure = lambda: self.on_market_cycle(message)`and pass the closure in the executor.submit method.

**Review by spyrostz — DISMISSED** [review]
One comment, LGTM once resolved

File: d3a_api_client/redis_aggregator.py (hunk 1 above)
Comment (Hassan754) [reply d3a_api_client/redis_aggregator.py:171]: Right , Thanks ! 
Adapted to your point

File: d3a_api_client/redis_aggregator.py [hunk 2]
```diff
@@ -167,51 +168,32 @@ def batch_command(self, batch_command_dict, is_blocking=True):
...
-
-        self.executor.submit(executor_function)
+        function_name = "on_tick"
+        function = lambda: self.on_tick(message)
+        self.executor.submit(execute_function_util, function=function, function_name=function_name,
+                             root_logger=root_logger)
 
     def _on_trade(self, message):
         logging.info(f"A trade took place on the device. Trade information: {message}")
+        function_name = "on_trade"
```
Comment (hannesdiedrich) [inline d3a_api_client/redis_aggregator.py:185]: you could save another line of code per function if you set the  `function_name` directly when calling `self.executor.submit`, right?

File: d3a_api_client/redis_aggregator.py [hunk 3]
```diff
@@ -167,51 +168,32 @@ def batch_command(self, batch_command_dict, is_blocking=True):
...
-        def executor_function():
-            try:
-                self.on_trade(message)
-            except Exception as e:
-                root_logger.error(
-                    f"_on_trade raised exception: {e}. \n Traceback: {traceback.format_exc()}")
-
-        self.executor.submit(executor_function)
+        self.executor.submit(execute_function_util, function=function, function_name=function_name,
+                             root_logger=root_logger)
```
Comment (hannesdiedrich) [inline d3a_api_client/redis_aggregator.py:189]: Is it really necessary to pass the root_logger here? How about running `root_logger = logging.getLogger()` in utils.py ?

**Review by hannesdiedrich — DISMISSED** [review]
Two minor comments, LGTM in general

File: d3a_api_client/utils.py [hunk 4]
```diff
@@ -210,4 +211,12 @@ def list_running_canary_networks_and_devices_with_live_data(domain_name):
         cn["uuid"]: cn["scenarioData"]["forecastStreamAreaMapping"]
         for cn in data["data"]["listCanaryNetworks"]["configurations"]
         if cn["resultsStatus"] == "running"
-    }
\ No newline at end of file
+    }
+
+
+def execute_function_util(function: callable, function_name):
```
Comment (fievelk) [inline d3a_api_client/utils.py:217]: Is the `function_name` needed? What if we remove it and use something like `function.__name__`?

Comment (Hassan754) [reply d3a_api_client/utils.py:217]: Unfortunately I'm passing a closure lambda function, not the original one so this is not possible

Comment (fievelk) [reply d3a_api_client/utils.py:217]: Edit: sorry, it's a lambda function. Ignore my previous comment :)

File: d3a_api_client/redis_aggregator.py [hunk 5]
```diff
@@ -167,51 +168,28 @@ def batch_command(self, batch_command_dict, is_blocking=True):
...
-            try:
-                self.on_market_cycle(message)
-            except Exception as e:
-                root_logger.error(
-                    f"on_market_cycle raised exception: {str(e)}. \n Traceback: {str(traceback.format_exc())}")
-
-        self.executor.submit(executor_function)
+        function = lambda: self.on_market_cycle(message)
+        self.executor.submit(execute_function_util, function=function, function_name="on_market_cycle",
+                             )
```
Comment (faizan2590) [inline d3a_api_client/redis_aggregator.py:173]: out of curiosity which editor are you using that suggests to place a closing parenthesis to a whole new line?

```suggestion
        self.executor.submit(execute_function_util, function=function, function_name="on_market_cycle")
```

File: d3a_api_client/redis_aggregator.py [hunk 6]
```diff
@@ -167,51 +168,28 @@ def batch_command(self, batch_command_dict, is_blocking=True):
...
-        def executor_function():
-            try:
-                self.on_tick(message)
-            except Exception as e:
-                root_logger.error(
-                    f"on_tick raised exception: {e}. \n Traceback: {traceback.format_exc()}")
-
-        self.executor.submit(executor_function)
+        function = lambda: self.on_tick(message)
+        self.executor.submit(execute_function_util, function=function, function_name="on_tick",
```
Comment (faizan2590) [inline d3a_api_client/redis_aggregator.py:178]: https://github.com/gridsingularity/d3a-api-client/pull/81/files#r563523127

File: d3a_api_client/redis_aggregator.py [hunk 7]
```diff
@@ -167,51 +168,28 @@ def batch_command(self, batch_command_dict, is_blocking=True):
...
 
-        def executor_function():
-            try:
-                self.on_trade(message)
-            except Exception as e:
-                root_logger.error(
-                    f"_on_trade raised exception: {e}. \n Traceback: {traceback.format_exc()}")
-
-        self.executor.submit(executor_function)
+        self.executor.submit(execute_function_util, function=function, function_name="on_trade",
```
Comment (faizan2590) [inline d3a_api_client/redis_aggregator.py:185]: https://github.com/gridsingularity/d3a-api-client/pull/81/files#r563523127

File: d3a_api_client/redis_aggregator.py [hunk 8]
```diff
@@ -167,51 +168,28 @@ def batch_command(self, batch_command_dict, is_blocking=True):
...
-        def executor_function():
-            try:
-                self.on_finish(message)
-            except Exception as e:
-                root_logger.error(
-                    f"on_finish raised exception: {e}. \n Traceback: {traceback.format_exc()}")
-
-        self.executor.submit(executor_function)
+        function = lambda: self.on_finish(message)
+        self.executor.submit(execute_function_util, function=function, function_name="on_finish",
```
Comment (faizan2590) [inline d3a_api_client/redis_aggregator.py:191]: https://github.com/gridsingularity/d3a-api-client/pull/81/files#r563523127

File: d3a_api_client/redis_client_base.py [hunk 9]
```diff
@@ -247,47 +251,30 @@ def _on_unregister(self, msg):
...
         logging.info(f"A new market was created. Market information: {message}")
-
-        def executor_function():
-            try:
-                self.on_market_cycle(message)
-            except Exception as e:
-                logging.error(f"_on_market_cycle raised exception: {e}. \n Traceback: {traceback.format_exc()}")
-        self.executor.submit(executor_function)
+        function = lambda: self.on_market_cycle(message)
+        self.executor.submit(execute_function_util, function=function, function_name="on_market_cycle",
```
Comment (faizan2590) [inline d3a_api_client/redis_client_base.py:255]: https://github.com/gridsingularity/d3a-api-client/pull/81/files#r563523127

File: d3a_api_client/redis_client_base.py [hunk 10]
```diff
@@ -247,47 +251,30 @@ def _on_unregister(self, msg):
...
-
-        def executor_function():
-            try:
-                self.on_tick(message)
-            except Exception as e:
-                logging.error(f"on_tick raised exception: {e}. \n Traceback: {traceback.format_exc()}")
-
-        self.executor.submit(executor_function)
+        function = lambda: self.on_tick(message)
+        self.executor.submit(execute_function_util, function=function, function_name="on_tick",
```
Comment (faizan2590) [inline d3a_api_client/redis_client_base.py:262]: https://github.com/gridsingularity/d3a-api-client/pull/81/files#r563523127

File: d3a_api_client/redis_aggregator.py (hunk 5 above)
Comment (Hassan754) [reply d3a_api_client/redis_aggregator.py:173]: This is not always case, maybe it was already on a new line after I deleted the extra argument

**Review by faizan2590 — APPROVED** [review]
One code formatting comment else LGTM

File: d3a_api_client/redis_aggregator.py [hunk 11]
```diff
@@ -167,51 +168,28 @@ def batch_command(self, batch_command_dict, is_blocking=True):
...
-            try:
-                self.on_market_cycle(message)
-            except Exception as e:
-                root_logger.error(
-                    f"on_market_cycle raised exception: {str(e)}. \n Traceback: {str(traceback.format_exc())}")
-
-        self.executor.submit(executor_function)
+        function = lambda: self.on_market_cycle(message)
+        self.executor.submit(execute_function_util, function=function,
+                             function_name="on_market_cycle",)
```
Comment (faizan2590) [inline d3a_api_client/redis_aggregator.py:173]: conforming with [PEP8 style guide in terms of when to use comma](https://www.python.org/dev/peps/pep-0008/#when-to-use-trailing-commas) there shouldn't be a trailing comma inside function arguments.

File: d3a_api_client/redis_aggregator.py [hunk 12]
```diff
@@ -167,51 +168,28 @@ def batch_command(self, batch_command_dict, is_blocking=True):
...
-            try:
-                self.on_tick(message)
-            except Exception as e:
-                root_logger.error(
-                    f"on_tick raised exception: {e}. \n Traceback: {traceback.format_exc()}")
-
-        self.executor.submit(executor_function)
+        function = lambda: self.on_tick(message)
+        self.executor.submit(execute_function_util, function=function,
+                             function_name="on_tick",)
```
Comment (faizan2590) [inline d3a_api_client/redis_aggregator.py:179]: https://github.com/gridsingularity/d3a-api-client/pull/81/files#r563539134

File: d3a_api_client/redis_aggregator.py [hunk 13]
```diff
@@ -167,51 +168,28 @@ def batch_command(self, batch_command_dict, is_blocking=True):
...
-        def executor_function():
-            try:
-                self.on_trade(message)
-            except Exception as e:
-                root_logger.error(
-                    f"_on_trade raised exception: {e}. \n Traceback: {traceback.format_exc()}")
-
-        self.executor.submit(executor_function)
+        self.executor.submit(execute_function_util, function=function,
+                             function_name="on_trade",)
```
Comment (faizan2590) [inline d3a_api_client/redis_aggregator.py:186]: https://github.com/gridsingularity/d3a-api-client/pull/81/files#r563539134

File: d3a_api_client/redis_aggregator.py [hunk 14]
```diff
@@ -167,51 +168,28 @@ def batch_command(self, batch_command_dict, is_blocking=True):
...
-            try:
-                self.on_finish(message)
-            except Exception as e:
-                root_logger.error(
-                    f"on_finish raised exception: {e}. \n Traceback: {traceback.format_exc()}")
-
-        self.executor.submit(executor_function)
+        function = lambda: self.on_finish(message)
+        self.executor.submit(execute_function_util, function=function,
+                             function_name="on_finish",)
```
Comment (faizan2590) [inline d3a_api_client/redis_aggregator.py:192]: https://github.com/gridsingularity/d3a-api-client/pull/81/files#r563539134

File: d3a_api_client/redis_aggregator.py (hunk 11 above)
Comment (Hassan754) [reply d3a_api_client/redis_aggregator.py]: Thanks for pointing at it, Done

**Review by faizan2590 — APPROVED** [review]

**Review by hannesdiedrich — APPROVED** [review]
LGTM

**Comment by Hassan754** [comment]
Merging this PR in 30 mins, still open for reviews

**Pull request merged**